readarr: init at 0.1.4.1596

jocelynthode opened the pull request:
###### Description of changes

This adds Readarr a Usenet/BitTorrent ebook/audiobook downloader. One of the many *arr tool

Closes https://github.com/NixOS/nixpkgs/issues/179941

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jocelynthode commented:
The package should build for `aarch64-linux` and `x86_64-darwin` but I cannot test these. I was able to build the application and run it `nix-build -A readarr ./` and `./result/bin/Readarr`

As I never added a module I'm not sure how to test it given that I'm using flakes in my configuration and I only found ways to test it when not using flakes. If someone can explain how to test it when using flakes, I can test it as well.

jocelynthode commented:
@purcell: Would you mind giving this PR a quick review please?

purcell commented:
Hey, `pkgs/servers/readarr/default.nix` looks good to me, I can't really comment on best practices for the tests and updating release notes etc.

SuperSandro2000 commented on nixos/modules/misc/ids.nix:
```diff
@@ -356,6 +356,7 @@ in
       rstudio-server = 324;
       localtimed = 325;
       automatic-timezoned = 326;
+      readarr = 327;
```
Please do not add new uuid unless absolutely necessary.

Did you follow a guide or wiki article we need to update?

jocelynthode replied:
I just looked at what was done for other *arr apps such as lidarr,sonarr,radarr etc. as they had their uids/gids here I did the same for the smoke-tests to work.

purcell replied:
@SuperSandro2000 is there a preferred approach? Happy to help migrate the other `*arr` apps if I know what to do.

SuperSandro2000 replied:
Just define a user and group under users.

Tests shouldn't require fixed uids and gids.

jocelynthode replied:
Done I have removed the fixed uids/gids :)

SuperSandro2000 commented on pkgs/servers/readarr/default.nix:
```diff
@@ -0,0 +1,53 @@
+    cp -r * $out/share/${pname}-${version}/.
+    makeWrapper "${dotnet-runtime}/bin/dotnet" $out/bin/Readarr \
+      --add-flags "$out/share/${pname}-${version}/Readarr.dll" \
+      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [
+        curl sqlite libmediainfo icu openssl ]}
```
```suggestion
      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ curl sqlite libmediainfo icu openssl ]}
```

SuperSandro2000 commented on pkgs/servers/readarr/default.nix:
```diff
@@ -0,0 +1,53 @@
+  version = "0.1.1.1320";
+
+  src = fetchurl {
+    url = "https://github.com/Readarr/Readarr/releases/download/v${version}/Readarr.develop.${version}.${os}-core-${arch}.tar.gz";
```
Why not build the dotnet project from source?

jocelynthode replied:
As before I followed was what done for the other *arr projects and they all downloaded the released archive.

I could try and see if we can succeed in building it from source. However I'm not that familiar with the dotnet environment. I would need some help in achieving this.

purcell replied:
I'm unaware of the background to building these from the archives. I remember that Jellyfin was also built like this at one point too, and in a7ccfa062cc it was updated to build from source. The current machinery for jellyfin might be a good example to copy for all the `*arr` servers in future, though I haven't tried it myself. If you felt like trying it for `readarr`, I could help switch the others over too. (I'm personally fine with the current approach in this PR in practice, so not trying to block it.)

jocelynthode replied:
I have tried to build it from source yesterday, but I was not able to do so. I had to write my own `fetch-deps` script as the default one does not care about the `NuGet.config` of the repository and was theon unable to generate the correct deps.nix

Afterwards I was still hit by a lot of errors related to wildcards in version name not being compatible with `Determinism` as I don't know anything about dotnet I gave up there.

@SuperSandro2000 I propose that we go forward with installing from the archive and we could have a second PR trying torefactor all the *arr modules to be built from source and correct the various things that you pointed at in this PR. I would not mind helping @purcell on this he he has more experience than me on dotnet.

SuperSandro2000 commented on pkgs/servers/readarr/default.nix:
```diff
@@ -0,0 +1,53 @@
+  meta = with lib; {
+    description = "A Usenet/BitTorrent ebook downloader";
+    homepage = "https://readarr.com";
+    license = licenses.gpl3;
```
If we do not build from source we need to set source [source provenance](https://github.com/NixOS/nixpkgs/blob/master/doc/stdenv/meta.chapter.md#source-provenance-sec-meta-sourceprovenance) instead,

jocelynthode replied:
I'll add it. I'd like to note however that other *arr package do not set this value. I think a global refactor of the various *arr packages would be in order to align with best practices.

Unfortunately I'm not sure I have the time to tackle the other packages.

purcell replied:
If you figure out source provenance for this one, I can look at the others.

jocelynthode replied:
I've used: `sourceProvenance = with lib.sourceTypes; [ binaryBytecode ];` for now.

SuperSandro2000 commented on nixos/modules/services/misc/readarr.nix:
```diff
@@ -0,0 +1,89 @@
+
+      dataDir = mkOption {
+        type = types.str;
+        default = "/var/lib/readarr/.config/Readarr";
```
```suggestion
        default = "/var/lib/readarr/";
```
This would be more like other services work

jocelynthode replied:
Again all the other *arr modules/packages in nixpkgs follow this standard. Should we change it for all of them?

purcell replied:
I imagine that might break existing installations, ie. they wouldn't find their config. So might be low value to fix up there. But if it works here with the shorter path, no reason not to use it.

jocelynthode replied:
I switched to @SuperSandro2000 suggestion as this is a new module. Let's go with good practices.

purcell commented:
> Anything in particular blocking this from getting merged?

I'm not aware of any blocker, and assume that it has languished due to the suggestion that we instead build from source. From my perspective, that might be preferable, but needn't hold it up because `readarr`/`sonarr` also do so.

etu commented:
@jocelynthode Hey, sorry I haven't come around to this sooner. Are you willing to rebase it to address the conflict?

I fully agree with the comment by @purcell.

jocelynthode committed:
readarr: init at 0.1.4.1596

jocelynthode commented:
@etu Thanks for taking a look at this. I've rebased and also updated to the latest version via the update script present :). I've also retested via `nix-build -A readarr ./` and running the resulting `Readarr` binary with success.
